### PR TITLE
JBIDE-25314 CDK launch configuration itests

### PR DESCRIPTION
**Pull request by odockal**
- New test class CDKLaunchConfigurationTest
	- CDKSmoke and CDKAllTests suites extended with new test class
	- CDKLaunchConfigurationDialog added into cdk.reddeer plugin
	- New mock files representing cdk binaries
	- Remove deprecated CDK3 test classes
	- Add check of OS and docker after restart

Signed-off-by: Ondrej Dockal <odockal@redhat.com>

# Pull Request Checklist
## General
* Is this a blocking issue or new feature? If yes, QE needs to +1 this PR

## Code
* Are method-/class-/variable-names meaningful?
* Are methods concise, not too long?
* Are catch blocks catching precise Exceptions only (no catch all)?

## Testing
* Are there unit-tests?
* Are there integration tests (or at least a jira to tackle these)?
* Is the non-happy path working, too?
* Are other parts that use the same component still working fine?

## Function
* Does it work?

### CODE REVIEW

**Comment by odockal** [comment]
testPR

**Comment by adietish** [comment]
testPR

**Comment by odockal** [comment]
testPR

**Comment by adietish** [comment]
@odockal i have to miss something. When launching the tests I see it complaining about a profile related error

```
Failed tests:
  CDK32ServerWizardTest.testNewCDK32ServerWizard:85->CDKServerWizardAbstractTest.assertSameMessage:144 Expected page description should contain text: is not compatible with this server adapter but has: A server adapter representing Red Hat Container Development Kit Version 3.2+
  CDK32ServerEditorTest.testInvalidMinishiftLocation:125->CDKServerEditorAbstractTest.checkEditorStateAfterSave:120->CDKServerEditorAbstractTest.verifyEditorCannotSave:178 Editor was saved successfully but exception was expected
  CDKLaunchConfigurationTest.testMinishiftHomePropagationIntoLaunchConfig:163 expected:<.../adietish/.minishift[Folder]> but was:<.../adietish/.minishift[]>
  CDK32ServerAdapterStartTest.testStartServerAdapter:44->CDKServerAdapterAbstractTest.startServerAdapter:135 Server's state went back to STOPPED
<terminated> Container Development Environment 3.2 [Red Hat CDK Server Startup Configuration] /Applications/cdk3/cdk-320
open /Users/adietish/.minishift/machines/minishift/config.json: no such file or directory

  CDK32ServerAdapterRestartTest.testCDERestart:46->CDKServerAdapterAbstractTest.startServerAdapter:135 Server's state went back to STOPPED
<terminated> Container Development Environment 3.2 [Red Hat CDK Server Startup Configuration] /Applications/cdk3/cdk-320
open /Users/adietish/.minishift/machines/minishift/config.json: no such file or directory

  CDK32ServerAdapterConnectionTest.testCDK3ServerAdapterConnection:43->CDKServerAdapterAbstractTest.startServerAdapter:135 Server's state went back to STOPPED
<terminated> Container Development Environment 3.2 [Red Hat CDK Server Startup Configuration] /Applications/cdk3/cdk-320
open /Users/adietish/.minishift/machines/minishift/config.json: no such file or directory


Tests run: 13, Failures: 6, Errors: 0, Skipped: 0
```
image

**Comment by adietish** [comment]
@robstryker dare to test this please?

**Comment by odockal** [comment]
@adietish The error in the pic suggests that there is something wrong with .minishift folder, perhaps try to delete old minishift config (minishift delete, remove .minishift and .kube...) and then minishift setup-cdk from fresh. I have never seen such error before.

**Comment by jkopriva** [comment]
@adietish for variables -Dminishift= and -Dminishift.profile= you must enter absolute path, otherwise the test will be failing, also for CDK32 tests you also need to set up -Dminishift= variable(something like: -Dminishift=/home/jkopriva/git/jkopriva/jbosstools-openshift/itests/org.jboss.tools.cdk.ui.bot.test/resources/cdk-files/linux/cdk-3.1.1-mock), @odockal is working on better variables handling

**Comment by robstryker** [comment]
When I run the tests, I get tons of failures. 

Tests run: 28, Failures: 12, Errors: 3, Skipped: 0

[rob@rawbdorable org.jboss.tools.cdk.ui.bot.test] (cdk-launch)$ mvn clean install -DskipTests=false -DskipITests=false -PITests -Dminishift.hypervisor=virtualbox -Dminishift=/home/rob/Downloads/cdk/minishift/cdk3.1.1.20170911/minishift  -Dminishift.profile=/home/rob/Downloads/cdk/minishift/cdk32a1/minishift -Ddevelopers.username=rstryker@redhat.com -Ddevelopers.password=xxxxx -Dsecurestorage.password=xxxxx -Dvagrantfile=/home/rob/Downloads/cdk/cdk-install/25-Jan-2017.rc0/cdk/components/rhel/rhel-ose/ -Dusage_reporting_enabled=false

One of the problems (for me) seems to be that the @RemoveCDKServers requirement doesn't seem to be running. If I change a number of dialog.finish(TimePeriod.MEDIUM);   with   dialog.cancel();
  in the following classes, I get much closer to a green build (and sometimes 100% success):  

CDK32ServerWizardTest.java
CDK3ServerWizardTest.java
CDKServerWizardTest.java


I've run these tests repeatedly for a day now and I can't really allocate any more time towards them. I also believe this patch needs to be rebased against master, because JBIDE-25308 will cause some of the tests to fail, and this was only committed recently.

**Comment by adietish** [comment]
@odockal re-installing cdk helped. But running the tests in the following way:
```
mvn clean verify -Dminishift.hypervisor=xhyve -Dminishift=/Applications/cdk3/cdk-320 -Dminishift.profile=/Applications/cdk3/cdk-320 -Ddevelopers.username=adietish@redhat.com -Ddevelopers.password=XXXXX -Dusage_reporting_enabled=false -PITests -Pcdk32-all-tests
```
I unfortunately now face the following errors:
```
Failed tests:
  CDK32ServerWizardTest.testNewCDK32ServerWizard:85->CDKServerWizardAbstractTest.assertSameMessage:144 Expected page description should contain text: is not compatible with this server adapter but has: A server adapter representing Red Hat Container Development Kit Version 3.2+
  CDK32ServerEditorTest.testInvalidMinishiftLocation:125->CDKServerEditorAbstractTest.checkEditorStateAfterSave:120->CDKServerEditorAbstractTest.verifyEditorCannotSave:178 Editor was saved successfully but exception was expected
  CDKLaunchConfigurationTest.testMinishiftHomePropagationIntoLaunchConfig:163 expected:<.../adietish/.minishift[Folder]> but was:<.../adietish/.minishift[]>

Tests run: 13, Failures: 3, Errors: 0, Skipped: 0
```

**Comment by adietish** [comment]
correcting the cmd-line to the following:
```
mvn clean verify -Dminishift.hypervisor=xhyve -Dminishift=/Applications/cdk3/cdk-311 -Dminishift.profile=/Applications/cdk3/cdk-320 -Ddevelopers.username=adietish@redhat.com -Ddevelopers.password=XXXXX -Dusage_reporting_enabled=false -PITests -Pcdk32-all-tests
```
I now have 1 error left:
```
Failed tests:
  CDKLaunchConfigurationTest.testMinishiftHomePropagationIntoLaunchConfig:163 expected:<.../adietish/.minishift[Folder]> but was:<.../adietish/.minishift[]>
```
image

**Comment by odockal** [comment]
@robstryker @adietish I thank you guys for the feedback! Today I have planned to incorporate your comments into code.

1. First thing,
```
Failed tests:
  CDKLaunchConfigurationTest.testMinishiftHomePropagationIntoLaunchConfig:163 expected:<.../adietish/.minishift[Folder]> but was:<.../adietish/.minishift[]>
```
this failure should be solved as https://github.com/jbosstools/jbosstools-openshift/pull/1606 was merged recently, so uage of newest devstudio should do the job. But I must say that I wrote the test before the PR was merged and thus I did not test it myself that it is working, will check today.

2. @robstryker @removecdkservers requirement is truly not running properly as there is a bug causing that requirement is not run in in @after part of test if test failed. (https://github.com/eclipse/reddeer/issues/1847) I will fix this today as well. (not the bug in reddeer but the cleaning in after part).

3. JBIDE-25308 have to check this. (including rebasing).

4. @robstryker Thing with dialog.finish(timeout) vs. dialog.cancel(). This is a huge pain. I would consider clicking finish button as a compulsory part of successful test case (wizard test, editor/adapter test), wouldn't it? Of course, It might be omitted in wizard testing as the same thing is being tested in two other test cases (editor/adapter), as you suggested, I will try this. 
To make it clear why this is happening: Problem here is that finish button pressing triggers 
"Refreshing server adapter list" job, this is one of the most favorite jobs to handle, it is not triggered always, takes non constant time to finish and is hard to kill :D Code waits for one minute giving the job time to finish itself, then it tries to kill the job, sometimes it does not work.

Thanks for patience with itests!

**Comment by odockal** [comment]
Do you have some newer devstudio or jbt? It is running for me with devstudio version 11.2.0-v20171115-0753-B1626.

**Comment by adietish** [comment]
i am running from master branch aka source plugins.

File: itests/org.jboss.tools.cdk.ui.bot.test/src/org/jboss/tools/cdk/ui/bot/test/server/editor/launch/CDKLaunchConfigurationTest.java [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+ *
+ */
+@RunWith(RedDeerSuite.class)
+public class CDKLaunchConfigurationTest extends CDKServerEditorAbstractTest {
+
+	private String hypervisor = MINISHIFT_HYPERVISOR;
+	
+	private CDKLaunchConfigurationDialog launchDialog; 
+	
+	private static final Logger log = Logger.getLogger(CDKLaunchConfigurationTest.class);
```
[the remark this answers is not in the document]
Comment (odockal) [reply itests/org.jboss.tools.cdk.ui.bot.test/src/org/jboss/tools/cdk/ui/bot/test/server/editor/launch/CDKLaunchConfigurationTest.java]: I believe that in this case we should always use lower case as we use "log" as a final reference instance of Logger class. It is always used to access Logger's method such as log or debug, so always in format log.info(...) or log.error(...). I like this convention more as you can type without using shift. And of course I found something on stackoverflow (https://stackoverflow.com/questions/1417190/should-a-static-final-logger-be-declared-in-upper-case) which support my individual preference. After all, all above is my feeling and I believe that usage of log in lower case is the most used convention across itests repos. but int the end, this is decision of contributors in given plugin. I would vote for using lower case for final reference variable that are used to access methods (thus always followed with a dot, please, see https://web.archive.org/web/20120911192801/developers.sun.com/sunstudio/products/archive/whitepapers/java-style.pdf - Field naming).

**Comment by adietish** [comment]
running the latest PR I have
```
Failed tests:
  CDKLaunchConfigurationTest.testMinishiftHomePropagationIntoLaunchConfig:163 expected:<.../adietish/.minishift[Folder]> but was:<.../adietish/.minishift[]>

Tests run: 13, Failures: 1, Errors: 0, Skipped: 0
```

**Comment by adietish** [comment]
@robstryker dare to run this and confirm/refute please?

**Comment by adietish** [comment]
i have this finally running fine, the problem being outdated maven artifacts. Killing .m2/org.jboss.tools.openshift and reinstalling them helped.

```
Results :

Tests run: 13, Failures: 0, Errors: 0, Skipped: 0
```

**Comment by adietish** [comment]
with latest master installed as maven artifacts, running this fails for me on 

```
Results :

Failed tests:
  CDKLaunchConfigurationTest.testServerNamePropagationIntoLaunchConfig:126 expected:<...ment Environment 3.2[x]> but was:<...ment Environment 3.2[]>

Tests run: 13, Failures: 1, Errors: 0, Skipped: 0
```

**Pull request merged**